engine/script/src/script_vmath.cpp

```diff
@@ -2299,6 +2299,166 @@ namespace dmScript
+    {
+        Matrix4 translation_matrix = Matrix4::identity();
+        const ScriptUserType translation_type = GetType(L, 1);
+
+        if (translation_type == SCRIPT_TYPE_VECTOR3)
+        {
+            Vector3* translation = CheckVector3(L, 1);
+            translation_matrix.setTranslation(*translation);
+        }
+        else if (translation_type == SCRIPT_TYPE_VECTOR4)
+        {
+            Vector4* translation = CheckVector4(L, 1);
+            translation_matrix.setTranslation(translation->getXYZ());
+        }
+        else
+        {
+            return luaL_error(L, "%s.%s accepts (%s|%s) as first argument.", SCRIPT_LIB_NAME, "matrix4_compose",
+                SCRIPT_TYPE_NAME_VECTOR3, SCRIPT_TYPE_NAME_VECTOR4);
+        }
+        Matrix4 rotation_matrix = Matrix4::rotation(*CheckQuat(L, 2));
+        Matrix4 scale_matrix = Matrix4::scale(*CheckVector3(L, 3));
+        Matrix4 result = translation_matrix * rotation_matrix * scale_matrix;
+
+        PushMatrix4(L, result);
+        return 1;
+    }
+
+    /*# creates a new matrix4 from scale vector
+     *
+     * Creates a new matrix constructed from scale vector
+     *
+     * @name vmath.matrix4_scale
+     * @param scale [type:vector3] scale
+     * @return matrix [type:matrix4] new matrix4
+     * @examples
+     *
+     * ```lua
+     * local scale = vmath.vector3(1, 0.5, 0.5)
+     * local result = vmath.matrix4_scale(scale)
+     * print(result) --> vmath.matrix4(1, 0, 0, 0, 0, 0.5, 0, 0, 0, 0, 0.5, 0, 0, 0, 0, 1)
+     * ```
+     */
+    static int Matrix4_Scale(lua_State* L)
+    {
+        Vector3* scale = CheckVector3(L, 1);
+        Matrix4 result = Matrix4::scale(*scale);
+        PushMatrix4(L, result);
+        return 1;
+    }
+
+    /*# clamp input value in range [min, max] and return clamped value
+     *
+     * Clamp input value to be in range of [min, max]. In case if input value has vector|vector3|vector4 type
+     * return new vector|vector3|vector4 with clamped value at every vector element.
+     *
+     * @name vmath.clamp
+     * @param value [type:number|vector3, vector4|vector] Input value or vector of values
+     * @param min [type:number] Min value border
+     * @param max [type:number] Max value border
+     * @return clamped_value [type:number|vector3|vector4|vector] Clamped value or vector
+     * @example
+     * 
+     * ```lua
+     * local value1 = 56
+	 * print(vmath.clamp(value1, 89, 134)) -> 89
+	 * local v1 = vmath.vector({ -45, 34, 13, 10, 55 })
+	 * print(vmath.clamp(v1, 0, 15)) -> vmath.vector (size: 5)
+	 * local v2 = vmath.vector3(190, 190, -10)
+	 * print(vmath.clamp(v2, -50, 150)) -> vmath.vector3(150, 150, -10)
+	 * local v3 = vmath.vector4(30, -30, 45, 1)
+	 * print(vmath.clamp(v3, 0, 20)) -> vmath.vector4(20, 0, 20, 1)
+     * ```
+     */
+    static int Vector_Clamp(lua_State* L)
+    {
+        float min = (float) luaL_checknumber(L, 2);
+        float max = (float) luaL_checknumber(L, 3);
```

Comment:
Potentially, we could also have a "per element" clamp.
I'm not sure what the actual user requirements are here tbh.
We can probably leave it until someone requests it.

Reply:
In that case I would suggest to have two separate functions:
* `vmath.clamp(v, min, max)` for number only
* `vmath.vclamp(vec, min_vec, max_vec)` for vectors only

WDYT?

Reply:
I prefer not to complicate it. vmath.clamp() for numbers and the elements of vectors will be very useful.